Auth protection improvements

lukasbestle opened the pull request.

lukasbestle committed:
More robust login failure tracking

All Exceptions that could be thrown while logging in are now first tracked and then passed on.

lukasbestle committed:
Only store first 50 chars of IP hash for privacy

lukasbestle committed:
Use cached user object in Auth::user()

lukasbestle committed:
Also apply auth protection to basic auth

lukasbestle committed:
Track login failures by email address as well

lukasbestle committed:
Automatically clean up .logins file

lukasbestle committed:
Clean up old .logins entries

bastianallgeier commented on config/api/routes/auth.php:
```diff
@@ -36,24 +36,21 @@
                     throw $e;
                 } else {
-                    // catch any kind of login error
+                    // hide the actual error to avoid leaking security-relevant information
```
You basically undid my changes from the last version here. I added the option to get proper info about failed logins when Kirby is in debug mode. There were quite a few people who had issues with failed logins and couldn't debug them properly. I know about the security risk, but I still think we need a way to have proper error handling with useful error messages when you actually want to debug this.

lukasbestle replied:
No, that feature is still there (see the three lines above). I have only cleaned up the code a bit – it still works exactly the same way.

bastianallgeier commented on i18n/translations/en.json:
```diff
@@ -169,7 +169,6 @@
   "error.user.password.notSame": "The passwords do not match",
   "error.user.password.undefined": "The user does not have a password",
   "error.user.role.invalid": "Please enter a valid role",
-  "error.user.undefined": "The user cannot be found",
```
I wouldn't remove this. This is also a useful message for the panel and I think we even use it somewhere else.

lukasbestle replied:
Before I removed it, I searched for `user.undefined` across the codebase and didn't find any remaining match. The error message is a duplicate of `error.user.notFound`.

bastianallgeier commented on src/Cms/Auth.php:
```diff
@@ -79,13 +80,11 @@ public function currentUserFromBasicAuth(BasicAuth $auth = null)
-            }
+        try {
+            return $this->validatePassword($auth->username(), $auth->password());
+        } catch (Throwable $e) {
```
Same as above. If we want to have a way to get proper error messages on failed logins, we need to let the exceptions bubble up.

lukasbestle replied:
That one was like this before as well. But I will see if I can improve on this further.

lukasbestle replied:
Fixed now (not yet pushed).

bastianallgeier commented on src/Cms/Auth.php:
```diff
@@ -158,24 +157,34 @@ public function impersonate(string $who = null)
+        $hash = hash('sha256', $this->kirby->visitor()->ip());
+
+        // only use the first 50 chars to ensure privacy
+        return substr($hash, 0, 50);
```
Good one!

bastianallgeier reviewed (changes requested):
I think we should add proper tests for this entire PR to avoid further problems with the login protection.

lukasbestle committed:
Fix API error handling on errors without route

lukasbestle committed:
Centralized auth error message handling

Hidden error details when debug mode is not enabled now apply to both session as well as basic auth.

lukasbestle committed:
Add Auth tests

lukasbestle committed:
Auth: Overwrite invalid log file automatically

lukasbestle committed:
Auth: Simpler isBlocked check, fix trials option

The timeout is now already checked when loading the file.

lukasbestle committed:
Ensure that $kirby->user() never throws

This would be a breaking change from previous behavior and would break a lot of other tests.

The detailed auth error can still be retrieved by calling $kirby->auth()->user() directly.

lukasbestle committed:
Auth: Cache Exception for multiple calls

Improves consistency of the method’s behavior because the result value itself is cached.

lukasbestle committed:
Fix influence on other tests

bastianallgeier reviewed (approved):
Looks great! Thanks for all the tests.
